File: bindings/nim/src/proven/safe_unit.nim
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
#
# Safe physical unit conversions.
# Thin wrapper over libproven FFI -- all logic lives in Idris.

import std/options
import lib_proven

proc convertLength*(value: float64, fromUnit, toUnit: LengthUnit): Option[float64] =
  ## Convert length between units.
  ## Returns None if the conversion fails.
  let res = provenUnitConvertLength(value, int32(ord(fromUnit)), int32(ord(toUnit)))
  if res.status == PROVEN_OK:
    return some(res.value)
  none(float64)

proc convertTemperature*(value: float64, fromUnit, toUnit: TempUnit): Option[float64] =
  ## Convert temperature between units.
  ## Returns None if the conversion fails.
  let res = provenUnitConvertTemp(value, int32(ord(fromUnit)), int32(ord(toUnit)))
  if res.status == PROVEN_OK:
    return some(res.value)
  none(float64)

File: bindings/php/src/Proven/SafeUnit.php
<?php
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// SafeUnit - FFI wrapper for proven_unit_* functions.
// All computation delegated to libproven; no logic reimplemented here.

declare(strict_types=1);

namespace Proven;

/**
 * Safe unit conversion via libproven FFI.
 *
 * Every method calls the corresponding libproven FFI function.
 * Unit enum values are integers defined by the libproven C ABI.
 */
final class SafeUnit
{
    /**
     * Convert a length value between units.
     *
     * @param float $value The value to convert.
     * @param int   $from  Source unit enum value.
     * @param int   $to    Target unit enum value.
     * @return float|null The converted value, or null on error.
     */
    public static function convertLength(float $value, int $from, int $to): ?float
    {
        $result = FFI::getLib()->proven_unit_convert_length($value, $from, $to);
        if ($result->status !== 0) {
            return null;
        }
        return (float) $result->value;
    }

    /**
     * Convert a temperature value between units.
     *
     * @param float $value The temperature to convert.
     * @param int   $from  Source unit enum value.
     * @param int   $to    Target unit enum value.
     * @return float|null The converted value, or null on error.
     */
    public static function convertTemp(float $value, int $from, int $to): ?float
    {
        $result = FFI::getLib()->proven_unit_convert_temp($value, $from, $to);
        if ($result->status !== 0) {
            return null;
        }
        return (float) $result->value;
    }
}
